In [1]:
%load_ext sql

In [3]:
%sql mysql://USERNAME:PASSWORD@HOST/DATABASE

# Exploratory queries
## 1.)
This query was performed make sure that there were not any null values inserted into the data. 
Null values could have meant there was an error either with the data collection or putting it in mysql which
would may have meant our data was not correct

In [7]:
%%sql 
SELECT *
FROM AccountsFollowed
WHERE following IS NULL;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
0 rows affected.


[]

From this query I learned that that were no null values inserted into the data meaning that we can move on and not reexamine our data collection

## 2.)
This query is done to see how many different users we have selected data from on what accounts they follow in addition to wayfair. This is done to insure that we are selected the proper amount of data from all users

In [20]:
%%sql 
SELECT COUNT(DISTINCT(username))
FROM AccountsFollowed;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(9,)]

This query shows that only 9 different users are in our dataset which will signfigantly affect the conclusions we can draw. 
If I was able to get the code working correctley it would have been much more users that I gathered data from. 

## 3.) 
This query was done to see the number of accounts following I have been able to get data where it is grouped by each 
Wayfair follower. This is done to verify if the amount of followers we are getting from each account makes sense

In [11]:
%%sql
SELECT 
    username, 
    COUNT(*) AS number_of_following
FROM AccountsFollowed
GROUP BY username;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
9 rows affected.


[('TracieR66729489', 9),
 ('mkmorani', 50),
 ('MolliesophiaLtd', 8),
 ('Louise_cole1', 50),
 ('xhaopengfei', 50),
 ('riaction20', 43),
 ('Arsalan36933021', 50),
 ('martybeaux', 50),
 ('ConvenConcepts', 20)]

The results show that there are a few accounts with far less accounts following than the others 
because due to rate limits I capped the amount of accounts following I am checking for each user at 50. This is likely due to the high variation of accounts people follow where some users follow thousands and others follow very few accounts

## 4.)
This query is seeing how many different accounts that people who follow wayfair are following that are in our dataset. THis shows wether or not we have a large enough number in the datset to draw good conclusions from

In [67]:
%%sql 
SELECT count(*) AS accounts_followed_count
FROM AccountsFollowed

 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(330,)]

This query showed that there are 330 total following accounts in our dataset that we can use to draw information from to inform our advertising decision. This query shows that there is simply not enough data and if I had figured out how to, ideally there would have been more. 

## 5.) 
In this query I am performing here is to checking on the wayfair followers table. I made a request to get the username of 150 wayfair followers and put them in a seperate table so the count should be 150. I did this to have an additional table where we could store more information in the future on each of these followers that follow wayfair. 

In [79]:
%%sql 

SELECT COUNT(DISTINCT(username)) AS count_of_users
FROM WayfairFollowers;


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(150,)]

From this query we can see that 150 wayfair followers we succesfuly put into the sql table menaing we can move forward from the data collection process

# Primary Question
My primary question is to find what are the most frequentley followed accounts by people that follow wayfair. In my project overview I am targeting a high value wayfair customer so this shows what accounts those high value customers also follow. This query is creating a view and joining the accounts that in the wayfair followers into the the accounts followed table. Thiw will make further queries easier and eliminate the need for joins

In [53]:
%%sql 
CREATE OR REPLACE VIEW accounts_following AS
    SELECT
        af.username,
        following
    FROM AccountsFollowed af
    LEFT JOIN WayfairFollowers wf
        ON af.username = wf.username;

 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
0 rows affected.


[]

There is not any information gained from this query, it is simply done to make the next two queries easier. 

# Secondary Queries
## 1.)
This query was performed to get the accounts that wayfair followers are following with counts greater than 1. I did greater than 1 becuase the dataset I am working with is so small. THis will show what other areas of interest a high potential wayfair customer has and where wayfair should target their advertising

In [60]:
%%sql
WITH rank_cte AS (
    Select 
        following, 
        count(*) AS follow_count, 
        RANK() OVER(
            ORDER BY COUNT(following) DESC
    ) AS account_rank
    FROM accounts_following
    GROUP BY following
    ORDER BY count(*) DESC
)
Select 
    following, 
    follow_count,
    account_rank
FROM rank_cte
WHERE follow_count > 1


 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
2 rows affected.


[('Wayfair', 8, 1), ('BBCNews', 2, 2)]

This query showed that besides wayfair, the only other account the two people follow among the 9 users is BBCnews, which even for a small dataset seems low. This shows that maybe wayfair should have ads on bbcnews or on their twitter page at least but we need a larger dataset for this query to be more beneficial

## 2.)
This secondary query is digging deeper into seeing what percentage of wayfair followers follow unqiue twitter accounts. This shows if the data will be useful to find commonalities and exploit those commonly followed twitter accounts for advertising tareting high potential wayfair customer.

In [65]:
%%sql
SELECT 
    ROUND(COUNT(DISTINCT(following))/COUNT(following) * 100, 2) AS unique_following_percent
FROM accounts_following;

 * mysql://admin:***@lmu-dev-01.cprqn4f2ofys.us-west-1.rds.amazonaws.com/WayfairTwitter
1 rows affected.


[(Decimal('97.58'),)]

The results of this show that most wayfair followers following unqiue twitter accounts so it is going to be hard to find commonly followed accounts in this dataset. This shows we are going to need to gather a lot of data to find commonly followed twitter accounts, something that I was not able to do given my expertise on gathering data

In [80]:
!pip install nbmerge

  Created wheel for nbmerge: filename=nbmerge-0.0.4-py2.py3-none-any.whl size=6413 sha256=26ed48ba57c7b64eb8be956a80ca1dd185c487214d027aca6cefd9c014f79aa5
  Stored in directory: c:\users\colem\appdata\local\pip\cache\wheels\0b\a8\31\4ae563f674b488c82ee72cfe0f86487c21532db97ae910b264
Successfully built nbmerge


In [85]:
!nbmerge TwitterAPIcollection.ipynb sql_analysis_cole_montoya.ipynb -o presentation.ipynb